x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/ShardFollowNodeTask.java

```diff
@@ -5,75 +5,323 @@
+        while (hasWriteBudget() && buffer.isEmpty() == false) {
+            Translog.Operation[] ops = new Translog.Operation[Math.min(params.getMaxWriteSize(), buffer.size())];
+            for (int i = 0; i < ops.length; i++) {
+                ops[i] = buffer.remove();
+            }
+            numConcurrentWrites++;
+            LOGGER.trace("{}[{}] write [{}/{}] [{}]", params.getFollowShardId(), numConcurrentWrites, ops[0].seqNo(),
+                ops[ops.length - 1].seqNo(), ops.length);
+            sendBulkShardOperationsRequest(ops);
+        }
+    }
+
+    private boolean hasWriteBudget() {
+        assert Thread.holdsLock(this);
+        if (numConcurrentWrites >= params.getMaxConcurrentWrites()) {
+            LOGGER.trace("{} maximum number of concurrent writes have been reached [{}]",
+                params.getFollowShardId(), numConcurrentWrites);
+            return false;
+        }
+        return true;
+    }
+
+
+    private void sendShardChangesRequest(long from, int maxOperationCount, long maxRequiredSeqNo) {
+        innerSendShardChangesRequest(from, maxOperationCount,
+            response -> {
+                retryCounter.set(0);
+                handleReadResponse(from, maxOperationCount, maxRequiredSeqNo, response);
+            },
+            e -> handleFailure(e, () -> sendShardChangesRequest(from, maxOperationCount, maxRequiredSeqNo)));
+    }
+
+    private void handleReadResponse(long from, int maxOperationCount, long maxRequiredSeqNo, ShardChangesAction.Response response) {
+        maybeUpdateMapping(response.getIndexMetadataVersion(), () -> {
+            synchronized (ShardFollowNodeTask.this) {
+                globalCheckpoint = Math.max(globalCheckpoint, response.getLeaderCheckpoint());
+                buffer.addAll(Arrays.asList(response.getOperations()));
+                coordinateWrites();
+
+                OptionalLong lastOp = Arrays.stream(response.getOperations()).mapToLong(Translog.Operation::seqNo).max();
+                if (lastOp.isPresent() && lastOp.getAsLong() < maxRequiredSeqNo) {
+                    long newFrom = lastOp.getAsLong() + 1;
+                    int newSize = (int) (maxRequiredSeqNo - lastOp.getAsLong());
+                    LOGGER.trace("{} received [{}] as last op while [{}] was expected, continue to read [{}/{}]...",
+                        params.getFollowShardId(), lastOp.getAsLong(), maxRequiredSeqNo, newFrom, maxOperationCount);
+                    sendShardChangesRequest(newFrom, newSize, maxRequiredSeqNo);
+                    return;
+                }
+
+                numConcurrentReads--;
+                if (response.getOperations().length != 0) {
+                    LOGGER.trace("{} post updating lastRequestedSeqno to [{}]", params.getFollowShardId(), lastRequestedSeqno);
+                    Translog.Operation firstOp = response.getOperations()[0];
+                    lastRequestedSeqno = Math.max(lastRequestedSeqno, lastOp.getAsLong());
```

Comment:
The global checkpoint field may already have been changed as that is the first thing what happens in this method. Then we need to keep track of the global checkpoint prior to changes it. You think it is worth doing that?

Reply:
I think it would have been worth it but now that you mention it - other requests may have changed this in the mean time too, so let's leave this assertion.